src/walker.test.ts

```diff
@@ -962,7 +962,48 @@ describe('flatten', () => {
       'd.e': 100,
       'd.f.0': 10,
       'd.f.1': 20,
-      'd.f.2': 30,
+      'd.f.2.g': 30,
+      'd.f.2.h.i': 40,
+    })
+  })
+  test('should flatten array', () => {
+    const arr = [10, 20, { a: { b: 20, c: 30, d: [40, { e: { f: 50 } }] } }]
```

Comment:
Did you explicitly copy the behavior from `flattenObjectsNotArrays` here? It may make more sense to leave the root array alone in this case.

Reply:
> It's a little weird to alter the return structure

I think it makes sense to return different types. For example `filterTree(isNotEmpty, {})` to return `undefined`.

Reply:
I reworked the logic here so that if an array is passed, the `objectsOnly` option is automatically set to true, and an array is returned.

Reply:
I reworked this again so that top-level arrays are flattened to an object of paths to values and top-level arrays with `objectsOnly` preserve all arrays.